crates/cranelift/src/func_environ.rs

```diff
@@ -1172,41 +1135,80 @@ impl<'a, 'func, 'module_env> Call<'a, 'func, 'module_env> {
         Ok(self.indirect_call_inst(sig_ref, func_addr, &real_call_args))
     }
 
-    /// Get the address of the call-indirect cache slot for a given callsite.
-    pub fn call_indirect_cache_slot_addr(
+    /// Get the address of the call-indirect cache tag and value slots
+    /// given a called table index.
+    pub fn call_indirect_cache_slot_tag_value_addr(
         &mut self,
-        call_site: CallIndirectSiteIndex,
+        table_cache: &TableCallCacheStyle,
         vmctx: ir::Value,
-    ) -> ir::Value {
-        let offset = self.env.offsets.vmctx_call_indirect_cache(call_site);
-        self.builder.ins().iadd_imm(vmctx, i64::from(offset))
+        index: ir::Value,
+    ) -> Option<(ir::Value, ir::Value)> {
+        match table_cache {
+            TableCallCacheStyle::None => None,
+            TableCallCacheStyle::DirectMapped { start, size_log2 } => {
+                // Compute the cache slot: `slot = start + (index & mask)`.
+                let masked_index = self
+                    .builder
+                    .ins()
+                    .band_imm(index, i64::try_from((1u64 << *size_log2) - 1).unwrap());
+                let offset_index = self
+                    .builder
+                    .ins()
+                    .iadd_imm(masked_index, i64::from(start.as_u32()));
+
+                // Compute addresses:
+                //   `p_tag = vmctx + tags_begin + sizeof(u32) * slot`
+                //   `p_value = vmctx + values_begin + sizeof(usize) * slot`
+                let tags_vmctx_offset = self.env.offsets.vmctx_call_indirect_cache_tags_begin();
+                let values_vmctx_offset = self.env.offsets.vmctx_call_indirect_cache_values_begin();
+                let tags_array = self
+                    .builder
+                    .ins()
+                    .iadd_imm(vmctx, i64::from(tags_vmctx_offset));
+                let values_array = self
+                    .builder
+                    .ins()
+                    .iadd_imm(vmctx, i64::from(values_vmctx_offset));
+                let tags_array_offset = self.builder.ins().imul_imm(
+                    offset_index,
+                    i64::try_from(std::mem::size_of::<u32>()).unwrap(),
+                );
+                let values_array_offset = self.builder.ins().imul_imm(
+                    offset_index,
+                    i64::try_from(std::mem::size_of::<usize>()).unwrap(),
```

Comment:
This should use `ptr_size` to handle cross-compiles between 32/64-bit (it's somewhere on `VMOffsets`, I forget which)

Reply:
Also could this and the above go ahead and use `ishl` instead of `imul_imm`?

Reply:
Done both!